Proposal for Synchronous Functionality

Jmeyer1292 opened the pull request:
This PR, not meant to be complete, is to demonstrate my proposal for the design of the synchronous portion of this library. It draws inspiration from ROS' concept of 'services' but does not have any dependencies on ROS.

In its current form, the library requires Cmake and uses [libsocket](https://github.com/dermesser/libsocket.git) for tcp communications. I expect this might change. You will need to build the above library to compile this project.

The library implements only a single command (the change program command) which I have tested and found to work correctly. 

The library currently does not:
 1. Handle errors
 2. Handle high-speed comms or similar async stuff
 3. Support the official keyence API
 4. Support USB
 5. Put much in source files

To implement a new command, one must create a file for the command (see change_program.h) and implement a class that follows a 'KeyenceCommandConcept':

1. Has two structures Response and Request 
2. Request has a typedef named ```response_type``` inside it naming the Request type
3. Response must have a function decode(...)
4. Request must have a function encode(...), a static int ```command_code``` and an int ```size``` indicating the size (in bytes) of the message data. 

Everything else is flexible.

See ```test_main.cpp``` for an example of how he library works currently.

Please give feedback.

gavanderhoorn commented:
In general this looks like a nice way to abstract msg (de)serialisation. It's not too complex, certainly not at the msg-specialisation-layer. Some msg specialisations will get rather complex though. We'll have to watch out for overly long `decode(..)` methods.

Some comments:

 1. stay away from C++11? This seems like a good driver to run on embedded platforms (sensor + small embedded board == ROS enabled keyence), C++11 support may be absent/incomplete. `socket++` is also using C++11, so that is something to investigate.
 1. keep endianness issues in mind (`insert()`, `extract()`)
 1. I like the explicit `Cient::send()` `Client::recv()`
 1. minor typo: recieve -> receive
 1. `Message::encode*Header(..)` static? Why not just functions in namespace?
 1. there is something to say for using pointers/unions directly with msg payloads: (de)serialisation very simple, no copying. It's not as clean as the current approach though perhaps.
 1. don't know if I like the `MutableBuffer operator+(..)` idea: is that for incrementing the buffer pointer? Syntax in `Client::sendRecieve(..)` is rather confusing.
 1. perhaps rename `MsgType::encode(MutableBuffer ..)` -> `MsgType::encodeInto(MutableBuffer ..)` (same for `MsgType::decode()`)
 1. need for `MutableBuffer` not entirely clear? `MutableBuffer` is i'face to `StreamBuffer`, but semantics allow client to iterate over internal data using pointer?
 1. why the asymmetry between request encoding and response decoding? Why does `Client::sendRecieve(..)` 'suddenly' decode the body? It doesn't encode it (at that layer) in the request.
 1. minor: `RequestHeader` is missing some fields (see dissector / protocol docs)

Jmeyer1292 commented:
@gavanderhoorn Thanks for taking a look at this PR!

1. I would tend to agree about c++11, but if we're targeting embedded platforms then we've got bigger issues. My current plan is to use exceptions for error reporting and that won't fly at all. Also heap allocation (and maybe virtual pointers).

2. I could add a #define like simple messages does.

3. Thanks.

4. +1

5. That's a good point, I will make those free functions.

6. I spent a while thinking about this one. I was worried about strict pointer aliasing issues. Plus I think the union method is technically undefined behavior because we would have more than one 'active' field. May not matter in reality though. If you think it would be a meaningful performance win, I will look into it.

7. (and 9) ```MutableBuffer``` is a meant to be a simple interface to any buffer with contiguous memory whether it's a c array or a class like vector.  I wanted the flexibility to change my buffer implementation, though it might be overkill with the current environment. As far as ```MutableBuffer operator+(..)```, I just wanted a cleaner way to increment the buffer pointer and reduce the buffer size. I found myself writing ```static_cast<char*>(void_ptr) + header_size, total_size - header_size). If you want I can toss the mutable buffer idea and go back to something closer to that. The command functions would get a void pointer and size instead of a MutableBuffer.

8. +1

9. See 7

10. I don't understand here. I am encoding and decoding in the same function.

11. I'm looking at the spreadsheet protocol documentation and all I see in the header for request is command_code and body_length plus a bunch of fixed fields. Maybe we're looking at different versions?

Jmeyer1292 commented:
I have updated the PR with:

1. Some general code cleanups.
2. Some basic error handling for broken socket connections, refused connections, etc...
3. Renamed functions
4. A single profile request command

Not implemented:

1. High speed mode (I'm waffling on how to do it)
2.  Detection of failure in the physical connection

One can induce a failure in physical connection by unplugging one end of the connection and the program will hang indefinitely until killed or the socket is reconnected. I see that in the simple messages library, synchronous sockets are used in conjunction with ```select``` to check for a timeout. Should I do a similar thing here in our ```send```/```recv``` functions or is that too much overhead for those calls?

shaun-edwards commented:
@Jmeyer1292, should we accept this and submit issues for missing functionality?  Is this PR a "working unit"?

Jmeyer1292 commented:
@shaun-edwards This is something I would like to have accepted and which I am willing to support and add more functionality. 

There are outstanding issues (most are mentioned in this thread), but I would to like to at least continue the review process on this unit.

Jmeyer1292 commented:
Ping

shaun-edwards commented:
+1 for experimental.  Let's address the issues as needed.  Thanks for the contribution!
